File: CSharpClient/Game/KS/Commands.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace KS.Commands
{
	public enum CommandMaterialType
	{
		Powder = 0,
		Steel = 1,
		Aluminum = 2,
		Shell = 3,
		Nickel = 4,
		Manganese = 5,
		Chromium = 6,
	}
	
	public enum CommandAmmoType
	{
		RifleBullet = 0,
		TankShell = 1,
		Bomb = 2,
		ArtilleryShell = 3,
		HMGBullet = 4,
	}
	
	public enum CommandAgentType
	{
		Warehouse = 0,
		Factory = 1,
	}
	
	public partial class BaseCommand : KSObject
	{
		public CommandAgentType? AgentType { get; set; }
		

		public BaseCommand()
		{
		}
		
		public new const string NameStatic = "BaseCommand";
		
		public override string Name() => "BaseCommand";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize AgentType
			s.Add((byte)((AgentType == null) ? 0 : 1));
			if (AgentType != null)
			{
				s.Add((byte)((sbyte)AgentType));
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize AgentType
			byte tmp0;
			tmp0 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp0 == 1)
			{
				sbyte tmp1;
				tmp1 = (sbyte)s[(int)offset];
				offset += sizeof(sbyte);
				AgentType = (CommandAgentType)tmp1;
			}
			else
				AgentType = null;
			
			return offset;
		}
	}
	
	public partial class Move : BaseCommand
	{
		public bool? Forward { get; set; }
		

		public Move()
		{
		}
		
		public new const string NameStatic = "Move";
		
		public override string Name() => "Move";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize parents
			s.AddRange(base.Serialize());
			
			// serialize Forward
			s.Add((byte)((Forward == null) ? 0 : 1));
			if (Forward != null)
			{
				s.AddRange(BitConverter.GetBytes((bool)Forward));
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize parents
			offset = base.Deserialize(s, offset);
			
			// deserialize Forward
			byte tmp2;
			tmp2 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp2 == 1)
			{
				Forward = BitConverter.ToBoolean(s, (int)offset);
				offset += sizeof(bool);
			}
			else
				Forward = null;
			
			return offset;
		}
	}
	
	public partial class PickMaterial : BaseCommand
	{
		public Dictionary<CommandMaterialType?, int?> Materials { get; set; }
		

		public PickMaterial()
		{
		}
		
		public new const string NameStatic = "PickMaterial";
		
		public override string Name() => "PickMaterial";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize parents
			s.AddRange(base.Serialize());
			
			// serialize Materials
			s.Add((byte)((Materials == null) ? 0 : 1));
			if (Materials != null)
			{
				List<byte> tmp3 = new List<byte>();
				tmp3.AddRange(BitConverter.GetBytes((uint)Materials.Count()));
				while (tmp3.Count > 0 && tmp3.Last() == 0)
					tmp3.RemoveAt(tmp3.Count - 1);
				s.Add((byte)tmp3.Count);
				s.AddRange(tmp3);
				
				foreach (var tmp4 in Materials)
				{
					s.Add((byte)((tmp4.Key == null) ? 0 : 1));
					if (tmp4.Key != null)
					{
						s.Add((byte)((sbyte)tmp4.Key));
					}
					
					s.Add((byte)((tmp4.Value == null) ? 0 : 1));
					if (tmp4.Value != null)
					{
						s.AddRange(BitConverter.GetBytes((int)tmp4.Value));
					}
				}
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize parents
			offset = base.Deserialize(s, offset);
			
			// deserialize Materials
			byte tmp5;
			tmp5 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp5 == 1)
			{
				byte tmp6;
				tmp6 = (byte)s[(int)offset];
				offset += sizeof(byte);
				byte[] tmp7 = new byte[sizeof(uint)];
				Array.Copy(s, offset, tmp7, 0, tmp6);
				offset += tmp6;
				uint tmp8;
				tmp8 = BitConverter.ToUInt32(tmp7, (int)0);
				
				Materials = new Dictionary<CommandMaterialType?, int?>();
				for (uint tmp9 = 0; tmp9 < tmp8; tmp9++)
				{
					CommandMaterialType? tmp10;
					byte tmp12;
					tmp12 = (byte)s[(int)offset];
					offset += sizeof(byte);
					if (tmp12 == 1)
					{
						sbyte tmp13;
						tmp13 = (sbyte)s[(int)offset];
						offset += sizeof(sbyte);
						tmp10 = (CommandMaterialType)tmp13;
					}
					else
						tmp10 = null;
					
					int? tmp11;
					byte tmp14;
					tmp14 = (byte)s[(int)offset];
					offset += sizeof(byte);
					if (tmp14 == 1)
					{
						tmp11 = BitConverter.ToInt32(s, (int)offset);
						offset += sizeof(int);
					}
					else
						tmp11 = null;
					
					Materials[tmp10] = tmp11;
				}
			}
			else
				Materials = null;
			
			return offset;
		}
	}
	
	public partial class PutMaterial : BaseCommand
	{
		public Dictionary<CommandMaterialType?, int?> Materials { get; set; }
		

		public PutMaterial()
		{
		}
		
		public new const string NameStatic = "PutMaterial";
		
		public override string Name() => "PutMaterial";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize parents
			s.AddRange(base.Serialize());
			
			// serialize Materials
			s.Add((byte)((Materials == null) ? 0 : 1));
			if (Materials != null)
			{
				List<byte> tmp15 = new List<byte>();
				tmp15.AddRange(BitConverter.GetBytes((uint)Materials.Count()));
				while (tmp15.Count > 0 && tmp15.Last() == 0)
					tmp15.RemoveAt(tmp15.Count - 1);
				s.Add((byte)tmp15.Count);
				s.AddRange(tmp15);
				
				foreach (var tmp16 in Materials)
				{
					s.Add((byte)((tmp16.Key == null) ? 0 : 1));
					if (tmp16.Key != null)
					{
						s.Add((byte)((sbyte)tmp16.Key));
					}
					
					s.Add((byte)((tmp16.Value == null) ? 0 : 1));
					if (tmp16.Value != null)
					{
						s.AddRange(BitConverter.GetBytes((int)tmp16.Value));
					}
				}
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize parents
			offset = base.Deserialize(s, offset);
			
			// deserialize Materials
			byte tmp17;
			tmp17 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp17 == 1)
			{
				byte tmp18;
				tmp18 = (byte)s[(int)offset];
				offset += sizeof(byte);
				byte[] tmp19 = new byte[sizeof(uint)];
				Array.Copy(s, offset, tmp19, 0, tmp18);
				offset += tmp18;
				uint tmp20;
				tmp20 = BitConverter.ToUInt32(tmp19, (int)0);
				
				Materials = new Dictionary<CommandMaterialType?, int?>();
				for (uint tmp21 = 0; tmp21 < tmp20; tmp21++)
				{
					CommandMaterialType? tmp22;
					byte tmp24;
					tmp24 = (byte)s[(int)offset];
					offset += sizeof(byte);
					if (tmp24 == 1)
					{
						sbyte tmp25;
						tmp25 = (sbyte)s[(int)offset];
						offset += sizeof(sbyte);
						tmp22 = (CommandMaterialType)tmp25;
					}
					else
						tmp22 = null;
					
					int? tmp23;
					byte tmp26;
					tmp26 = (byte)s[(int)offset];
					offset += sizeof(byte);
					if (tmp26 == 1)
					{
						tmp23 = BitConverter.ToInt32(s, (int)offset);
						offset += sizeof(int);
					}
					else
						tmp23 = null;
					
					Materials[tmp22] = tmp23;
				}
			}
			else
				Materials = null;
			
			return offset;
		}
	}
	
	public partial class PickAmmo : BaseCommand
	{
		public Dictionary<CommandAmmoType?, int?> Ammos { get; set; }
		

		public PickAmmo()
		{
		}
		
		public new const string NameStatic = "PickAmmo";
		
		public override string Name() => "PickAmmo";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize parents
			s.AddRange(base.Serialize());
			
			// serialize Ammos
			s.Add((byte)((Ammos == null) ? 0 : 1));
			if (Ammos != null)
			{
				List<byte> tmp27 = new List<byte>();
				tmp27.AddRange(BitConverter.GetBytes((uint)Ammos.Count()));
				while (tmp27.Count > 0 && tmp27.Last() == 0)
					tmp27.RemoveAt(tmp27.Count - 1);
				s.Add((byte)tmp27.Count);
				s.AddRange(tmp27);
				
				foreach (var tmp28 in Ammos)
				{
					s.Add((byte)((tmp28.Key == null) ? 0 : 1));
					if (tmp28.Key != null)
					{
						s.Add((byte)((sbyte)tmp28.Key));
					}
					
					s.Add((byte)((tmp28.Value == null) ? 0 : 1));
					if (tmp28.Value != null)
					{
						s.AddRange(BitConverter.GetBytes((int)tmp28.Value));
					}
				}
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize parents
			offset = base.Deserialize(s, offset);
			
			// deserialize Ammos
			byte tmp29;
			tmp29 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp29 == 1)
			{
				byte tmp30;
				tmp30 = (byte)s[(int)offset];
				offset += sizeof(byte);
				byte[] tmp31 = new byte[sizeof(uint)];
				Array.Copy(s, offset, tmp31, 0, tmp30);
				offset += tmp30;
				uint tmp32;
				tmp32 = BitConverter.ToUInt32(tmp31, (int)0);
				
				Ammos = new Dictionary<CommandAmmoType?, int?>();
				for (uint tmp33 = 0; tmp33 < tmp32; tmp33++)
				{
					CommandAmmoType? tmp34;
					byte tmp36;
					tmp36 = (byte)s[(int)offset];
					offset += sizeof(byte);
					if (tmp36 == 1)
					{
						sbyte tmp37;
						tmp37 = (sbyte)s[(int)offset];
						offset += sizeof(sbyte);
						tmp34 = (CommandAmmoType)tmp37;
					}
					else
						tmp34 = null;
					
					int? tmp35;
					byte tmp38;
					tmp38 = (byte)s[(int)offset];
					offset += sizeof(byte);
					if (tmp38 == 1)
					{
						tmp35 = BitConverter.ToInt32(s, (int)offset);
						offset += sizeof(int);
					}
					else
						tmp35 = null;
					
					Ammos[tmp34] = tmp35;
				}
			}
			else
				Ammos = null;
			
			return offset;
		}
	}
	
	public partial class PutAmmo : BaseCommand
	{
		

		public PutAmmo()
		{
		}
		
		public new const string NameStatic = "PutAmmo";
		
		public override string Name() => "PutAmmo";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize parents
			s.AddRange(base.Serialize());
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize parents
			offset = base.Deserialize(s, offset);
			
			return offset;
		}
	}
} // namespace KS.Commands

File: PythonClient/ks/commands.py
# -*- coding: utf-8 -*-

# python imports
import sys
import struct
from enum import Enum

PY3 = sys.version_info > (3,)


class CommandMaterialType(Enum):
	Powder = 0
	Steel = 1
	Aluminum = 2
	Shell = 3
	Nickel = 4
	Manganese = 5
	Chromium = 6


class CommandAmmoType(Enum):
	RifleBullet = 0
	TankShell = 1
	Bomb = 2
	ArtilleryShell = 3
	HMGBullet = 4


class CommandAgentType(Enum):
	Warehouse = 0
	Factory = 1


class BaseCommand(object):

	@staticmethod
	def name():
		return 'BaseCommand'


	def __init__(self, agent_type=None):
		self.initialize(agent_type)
	

	def initialize(self, agent_type=None):
		self.agent_type = agent_type
	

	def serialize(self):
		s = b''
		
		# serialize self.agent_type
		s += b'\x00' if self.agent_type is None else b'\x01'
		if self.agent_type is not None:
			s += struct.pack('b', self.agent_type.value)
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize self.agent_type
		tmp0 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp0:
			tmp1 = struct.unpack('b', s[offset:offset + 1])[0]
			offset += 1
			self.agent_type = CommandAgentType(tmp1)
		else:
			self.agent_type = None
		
		return offset


class Move(BaseCommand):

	@staticmethod
	def name():
		return 'Move'


	def __init__(self, agent_type=None, forward=None):
		self.initialize(agent_type, forward)
	

	def initialize(self, agent_type=None, forward=None):
		BaseCommand.initialize(self, agent_type)
		
		self.forward = forward
	

	def serialize(self):
		s = b''
		
		# serialize parents
		s += BaseCommand.serialize(self)
		
		# serialize self.forward
		s += b'\x00' if self.forward is None else b'\x01'
		if self.forward is not None:
			s += struct.pack('?', self.forward)
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize parents
		offset = BaseCommand.deserialize(self, s, offset)
		
		# deserialize self.forward
		tmp2 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp2:
			self.forward = struct.unpack('?', s[offset:offset + 1])[0]
			offset += 1
		else:
			self.forward = None
		
		return offset


class PickMaterial(BaseCommand):

	@staticmethod
	def name():
		return 'PickMaterial'


	def __init__(self, agent_type=None, materials=None):
		self.initialize(agent_type, materials)
	

	def initialize(self, agent_type=None, materials=None):
		BaseCommand.initialize(self, agent_type)
		
		self.materials = materials
	

	def serialize(self):
		s = b''
		
		# serialize parents
		s += BaseCommand.serialize(self)
		
		# serialize self.materials
		s += b'\x00' if self.materials is None else b'\x01'
		if self.materials is not None:
			tmp3 = b''
			tmp3 += struct.pack('I', len(self.materials))
			while len(tmp3) and tmp3[-1] == b'\x00'[0]:
				tmp3 = tmp3[:-1]
			s += struct.pack('B', len(tmp3))
			s += tmp3
			
			for tmp4 in self.materials:
				s += b'\x00' if tmp4 is None else b'\x01'
				if tmp4 is not None:
					s += struct.pack('b', tmp4.value)
				s += b'\x00' if self.materials[tmp4] is None else b'\x01'
				if self.materials[tmp4] is not None:
					s += struct.pack('i', self.materials[tmp4])
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize parents
		offset = BaseCommand.deserialize(self, s, offset)
		
		# deserialize self.materials
		tmp5 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp5:
			tmp6 = struct.unpack('B', s[offset:offset + 1])[0]
			offset += 1
			tmp7 = s[offset:offset + tmp6]
			offset += tmp6
			tmp7 += b'\x00' * (4 - tmp6)
			tmp8 = struct.unpack('I', tmp7)[0]
			
			self.materials = {}
			for tmp9 in range(tmp8):
				tmp12 = struct.unpack('B', s[offset:offset + 1])[0]
				offset += 1
				if tmp12:
					tmp13 = struct.unpack('b', s[offset:offset + 1])[0]
					offset += 1
					tmp10 = CommandMaterialType(tmp13)
				else:
					tmp10 = None
				tmp14 = struct.unpack('B', s[offset:offset + 1])[0]
				offset += 1
				if tmp14:
					tmp11 = struct.unpack('i', s[offset:offset + 4])[0]
					offset += 4
				else:
					tmp11 = None
				self.materials[tmp10] = tmp11
		else:
			self.materials = None
		
		return offset


class PutMaterial(BaseCommand):

	@staticmethod
	def name():
		return 'PutMaterial'


	def __init__(self, agent_type=None, materials=None):
		self.initialize(agent_type, materials)
	

	def initialize(self, agent_type=None, materials=None):
		BaseCommand.initialize(self, agent_type)
		
		self.materials = materials
	

	def serialize(self):
		s = b''
		
		# serialize parents
		s += BaseCommand.serialize(self)
		
		# serialize self.materials
		s += b'\x00' if self.materials is None else b'\x01'
		if self.materials is not None:
			tmp15 = b''
			tmp15 += struct.pack('I', len(self.materials))
			while len(tmp15) and tmp15[-1] == b'\x00'[0]:
				tmp15 = tmp15[:-1]
			s += struct.pack('B', len(tmp15))
			s += tmp15
			
			for tmp16 in self.materials:
				s += b'\x00' if tmp16 is None else b'\x01'
				if tmp16 is not None:
					s += struct.pack('b', tmp16.value)
				s += b'\x00' if self.materials[tmp16] is None else b'\x01'
				if self.materials[tmp16] is not None:
					s += struct.pack('i', self.materials[tmp16])
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize parents
		offset = BaseCommand.deserialize(self, s, offset)
		
		# deserialize self.materials
		tmp17 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp17:
			tmp18 = struct.unpack('B', s[offset:offset + 1])[0]
			offset += 1
			tmp19 = s[offset:offset + tmp18]
			offset += tmp18
			tmp19 += b'\x00' * (4 - tmp18)
			tmp20 = struct.unpack('I', tmp19)[0]
			
			self.materials = {}
			for tmp21 in range(tmp20):
				tmp24 = struct.unpack('B', s[offset:offset + 1])[0]
				offset += 1
				if tmp24:
					tmp25 = struct.unpack('b', s[offset:offset + 1])[0]
					offset += 1
					tmp22 = CommandMaterialType(tmp25)
				else:
					tmp22 = None
				tmp26 = struct.unpack('B', s[offset:offset + 1])[0]
				offset += 1
				if tmp26:
					tmp23 = struct.unpack('i', s[offset:offset + 4])[0]
					offset += 4
				else:
					tmp23 = None
				self.materials[tmp22] = tmp23
		else:
			self.materials = None
		
		return offset


class PickAmmo(BaseCommand):

	@staticmethod
	def name():
		return 'PickAmmo'


	def __init__(self, agent_type=None, ammos=None):
		self.initialize(agent_type, ammos)
	

	def initialize(self, agent_type=None, ammos=None):
		BaseCommand.initialize(self, agent_type)
		
		self.ammos = ammos
	

	def serialize(self):
		s = b''
		
		# serialize parents
		s += BaseCommand.serialize(self)
		
		# serialize self.ammos
		s += b'\x00' if self.ammos is None else b'\x01'
		if self.ammos is not None:
			tmp27 = b''
			tmp27 += struct.pack('I', len(self.ammos))
			while len(tmp27) and tmp27[-1] == b'\x00'[0]:
				tmp27 = tmp27[:-1]
			s += struct.pack('B', len(tmp27))
			s += tmp27
			
			for tmp28 in self.ammos:
				s += b'\x00' if tmp28 is None else b'\x01'
				if tmp28 is not None:
					s += struct.pack('b', tmp28.value)
				s += b'\x00' if self.ammos[tmp28] is None else b'\x01'
				if self.ammos[tmp28] is not None:
					s += struct.pack('i', self.ammos[tmp28])
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize parents
		offset = BaseCommand.deserialize(self, s, offset)
		
		# deserialize self.ammos
		tmp29 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp29:
			tmp30 = struct.unpack('B', s[offset:offset + 1])[0]
			offset += 1
			tmp31 = s[offset:offset + tmp30]
			offset += tmp30
			tmp31 += b'\x00' * (4 - tmp30)
			tmp32 = struct.unpack('I', tmp31)[0]
			
			self.ammos = {}
			for tmp33 in range(tmp32):
				tmp36 = struct.unpack('B', s[offset:offset + 1])[0]
				offset += 1
				if tmp36:
					tmp37 = struct.unpack('b', s[offset:offset + 1])[0]
					offset += 1
					tmp34 = CommandAmmoType(tmp37)
				else:
					tmp34 = None
				tmp38 = struct.unpack('B', s[offset:offset + 1])[0]
				offset += 1
				if tmp38:
					tmp35 = struct.unpack('i', s[offset:offset + 4])[0]
					offset += 4
				else:
					tmp35 = None
				self.ammos[tmp34] = tmp35
		else:
			self.ammos = None
		
		return offset


class PutAmmo(BaseCommand):

	@staticmethod
	def name():
		return 'PutAmmo'


	def __init__(self, agent_type=None):
		self.initialize(agent_type)
	

	def initialize(self, agent_type=None):
		BaseCommand.initialize(self, agent_type)
	

	def serialize(self):
		s = b''
		
		# serialize parents
		s += BaseCommand.serialize(self)
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize parents
		offset = BaseCommand.deserialize(self, s, offset)
		
		return offset
